### PR TITLE
[Feature] Support training and testing on HuMMan dataset

### PR DESCRIPTION
**Pull request by caizhongang**
To merge after #73 .

- [x] Add support for loading images from .smc files
- [x] Add support for iPhone image rotation and associated intrinsic/extrinsic transformation in SMCReader
- [x] Add unit tests

### CODE REVIEW

File: mmhuman3d/data/datasets/pipelines/loading.py [hunk 1]
```diff
@@ -43,8 +44,19 @@ def __call__(self, results):
...
+                                          'but image_id is not provided.'
+            device, device_id, frame_id = results['image_id']
+            smc_reader = SMCReader(filename)
+            img = smc_reader.get_color(
+                device, device_id, frame_id, disable_tqdm=True)
+            img = img.squeeze()  # (1, H, W, 3) -> (H, W, 3)
+            del smc_reader
+        else:
+            img_bytes = self.file_client.get(filename)
+            img = mmcv.imfrombytes(img_bytes, flag=self.color_type)
```
Comment (yl-1993) [inline mmhuman3d/data/datasets/pipelines/loading.py:58]: The default `channel_order` is `bgr` for [`mmcv.imfrombytes()` ](https://mmcv.readthedocs.io/en/latest/_modules/mmcv/image/io.html#imfrombytes), while `smc_reader.get_color()` returns `rgb` image, will this cause some potential problem?

File: mmhuman3d/data/datasets/human_image_dataset.py [hunk 2]
```diff
@@ -109,6 +109,10 @@ def prepare_raw_data(self, idx: int):
         image_path = self.human_data['image_path'][idx]
         info['image_path'] = os.path.join(self.data_prefix, 'datasets',
                                           self.dataset_name, image_path)
+        if image_path.endswith('smc'):
```
Comment (yl-1993) [inline mmhuman3d/data/datasets/human_image_dataset.py:112]: We may add a test for this case in [tests/test_datasets/test_human_image_dataset.py](https://github.com/open-mmlab/mmhuman3d/blob/main/tests/test_datasets/test_human_image_dataset.py).

File: tests/test_datasets/test_pipelines.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+def test_load_image_from_file_smc():
+    results = {'img_prefix': None, 'image_path': test_smc_path}
+
+    pipeline = LoadImageFromFile()
+
+    with pytest.raises(AssertionError):
+        results = pipeline(results)
+
+    results['image_id'] = ('Kinect', 0, 0)
```
Comment (yl-1993) [inline tests/test_datasets/test_pipelines.py:31]: Do we need to also test `('iPhone', 0, 0)`?

File: tests/test_datasets/test_pipelines.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    results = pipeline(results)
+
+    assert results['filename'] == results['ori_filename'] == test_smc_path
+    assert isinstance(results['img'], np.ndarray)
+    assert results['img_shape'] == results['ori_shape'] == (1080, 1920)
```
Comment (yl-1993) [inline tests/test_datasets/test_pipelines.py:36]: The difference between L19 and L36 reminds me of a question, have we documented the order of `ori_shape` is `(H, W)` somewhere?

File: mmhuman3d/data/datasets/pipelines/loading.py (hunk 1 above)
Comment (caizhongang) [reply mmhuman3d/data/datasets/pipelines/loading.py:60]: Fixed in https://github.com/open-mmlab/mmhuman3d/pull/76/commits/b518ac091630c97fafe8a045ef2b76566f8342af.

File: tests/test_datasets/test_pipelines.py (hunk 4 above)
Comment (caizhongang) [reply tests/test_datasets/test_pipelines.py:36]: Fixed in 8c9bd6e22491fd9ff9383028f544381460c4b8aa.

File: tests/test_datasets/test_pipelines.py (hunk 3 above)
Comment (caizhongang) [reply tests/test_datasets/test_pipelines.py:31]: Fixed in 8c9bd6e22491fd9ff9383028f544381460c4b8aa.

File: tests/test_datasets/test_pipelines.py (hunk 4 above)
Comment (caizhongang) [reply tests/test_datasets/test_pipelines.py:36]: I don't think so. I have add it to #54 .

File: mmhuman3d/data/datasets/human_image_dataset.py (hunk 2 above)
Comment (caizhongang) [reply mmhuman3d/data/datasets/human_image_dataset.py:112]: Added in e67820f98c0353290bd899b3a29516b30d44d7b6.

**Comment by yl-1993** [comment]
Looks good to me! Could you please help take a look? @kimren227

File: mmhuman3d/data/data_structures/smc_reader.py [hunk 5]
```diff
@@ -222,22 +245,58 @@ def get_iphone_extrinsics(self, iphone_id=0, homogeneous=True):
...
+        extrinsics = np.identity(4, dtype=float)
+        extrinsics[:3, :3] = R
+        extrinsics[:3, 3] = T
+
+        if vertical:
+            # 90-degree clockwise rotation around z-axis
+            R = np.eye(4)
+            R[:2, :2] = np.array([[0, -1], [1, 0]])
+            # Note the extrinsics is cam2world
+            extrinsics = extrinsics @ np.linalg.inv(R)
```
Comment (kimren227) [inline mmhuman3d/data/data_structures/smc_reader.py:263]: Have we tested this extrinsic?

File: mmhuman3d/data/data_structures/smc_reader.py [hunk 6]
```diff
@@ -382,32 +444,38 @@ def get_iphone_color(self, iphone_id=0, frame_id=None, disable_tqdm=False):
...
-            for i in tqdm.tqdm(frame_list, disable=disable_tqdm):
-                frames.append(
-                    self.__read_color_from_bytes__(
-                        self.smc['iPhone'][str(iphone_id)]['Color'][str(i)][(
-                        )]))
-            return np.stack(frames, axis=0)
-        else:
+        elif isinstance(frame_id, list):
+            frame_list = frame_id
+        elif isinstance(frame_id, int):
```
Comment (kimren227) [inline mmhuman3d/data/data_structures/smc_reader.py:462]: Do we have an else statement to capture invalid inputs?

File: mmhuman3d/data/data_structures/smc_reader.py [hunk 7]
```diff
@@ -420,27 +488,31 @@ def get_iphone_depth(self, iphone_id=0, frame_id=None, disable_tqdm=False):
...
-                    self.smc['iPhone'][str(iphone_id)]['Depth'][str(i)][()])
-            return np.stack(frames, axis=0)
-        else:
-            assert frame_id < self.iphone_num_frames,\
+        frames = []
+        if frame_id is None:
+            frame_list = range(self.get_iphone_num_frames())
+        elif isinstance(frame_id, list):
+            frame_list = frame_id
+        elif isinstance(frame_id, int):
```
Comment (kimren227) [inline mmhuman3d/data/data_structures/smc_reader.py:506]: Do we have an else statement to capture invalid inputs?

File: tests/test_smc_reader.py [hunk 8]
```diff
@@ -108,46 +91,82 @@ def test_get_iphone_intrinsics():
...
         _ = smc.get_color(device='Kinect', device_id=10, frame_id=0)
 
-    with pytest.raises(KeyError):
-        _ = smc.get_color(device='iPhone', device_id=1, frame_id=0)
+    kinect_color = smc.get_color(device='Kinect', device_id=0, frame_id=0)
+    assert kinect_color.shape == (1, 1080, 1920, 3), \
+        'Kinect Color should have resolution of 1x1080x1920x3'
+
+
+def test_get_kinect_depth():
```
Comment (kimren227) [inline tests/test_smc_reader.py:156]: Maybe rename this test?

File: tests/test_smc_reader.py [hunk 9]
```diff
@@ -375,3 +394,51 @@ def test_get_smpl_by_frame():
     assert isinstance(body_pose, np.ndarray)
     assert isinstance(transl, np.ndarray)
     assert isinstance(betas, np.ndarray)
+
+
+def test_iphone_rotation():
```
Comment (kimren227) [inline tests/test_smc_reader.py:399]: Nice test!

**Review by kimren227 — COMMENTED** [review]
Looks good to me, just some small issues.

File: mmhuman3d/data/data_structures/smc_reader.py [hunk 10]
```diff
@@ -201,9 +208,21 @@ def get_iphone_intrinsics(self, iphone_id=0, frame_id=0):
             frame_id)]
         camera_info = json.loads(camera_info[()])
         intrinsics = np.asarray(camera_info['cameraIntrinsics']).transpose()
+
+        if vertical:
```
[the remark this answers is not in the document]
Comment (caizhongang) [reply mmhuman3d/data/data_structures/smc_reader.py:215]: Added in https://github.com/open-mmlab/mmhuman3d/pull/76/commits/111db02d4155d34a1aa8b25c0dcd348ca73f5e05.

File: mmhuman3d/data/data_structures/smc_reader.py (hunk 5 above)
Comment (caizhongang) [reply mmhuman3d/data/data_structures/smc_reader.py:270]: It is included in the [`test_iphone_rotation()` ](https://github.com/open-mmlab/mmhuman3d/blob/feature_load_smc/tests/test_smc_reader.py#L428-L431).

File: mmhuman3d/data/data_structures/smc_reader.py (hunk 6 above)
Comment (caizhongang) [reply mmhuman3d/data/data_structures/smc_reader.py:471]: Fixed in https://github.com/open-mmlab/mmhuman3d/pull/76/commits/4d5dddf207b9419fd6ef5e266b1fb6f7dbf3cbb8.

File: mmhuman3d/data/data_structures/smc_reader.py (hunk 7 above)
Comment (caizhongang) [reply mmhuman3d/data/data_structures/smc_reader.py:517]: Fixed in https://github.com/open-mmlab/mmhuman3d/pull/76/commits/4d5dddf207b9419fd6ef5e266b1fb6f7dbf3cbb8.

File: tests/test_smc_reader.py (hunk 8 above)
Comment (caizhongang) [reply tests/test_smc_reader.py]: Rearranged in https://github.com/open-mmlab/mmhuman3d/pull/76/commits/93bf306e0e9ea77c01a8d7d349f79f1e67dd53ae.

**Review by kimren227 — APPROVED** [review]

**Pull request merged**